openff/toolkit/topology/topology.py

```diff
@@ -1079,15 +1099,19 @@ def identical_molecule_groups(self):
 
         {0: [[0, {0: 0, 1: 1, 2: 2}], [1, {0: 1, 1: 0, 2: 2}]]}
         """
-        identity_maps = self._identify_chemically_identical_molecules()
+        # Check whether this was run previously, and a cached result is available.
+        if self._cached_chemically_identical_molecules is not None:
+            return self._cached_chemically_identical_molecules
+
         # Convert molecule identity maps into groups of identical molecules
-        groupings = {}
-        for molecule_idx in identity_maps.keys():
-            unique_mol, atom_map = identity_maps[molecule_idx]
-            groupings[unique_mol] = groupings.get(unique_mol, list()) + [
-                [molecule_idx, atom_map]
-            ]
-        return groupings
+        identity_maps = self._identify_chemically_identical_molecules()
+        groupings = defaultdict(list)
+        for molecule_idx, (unique_mol, atom_map) in identity_maps.items():
+            groupings[unique_mol] += [[molecule_idx, atom_map]]
+
+        self._cached_chemically_identical_molecules = groupings
+
+        return self._cached_chemically_identical_molecules
 
     def _identify_chemically_identical_molecules(self):
```

Comment:
It's in the docstring but it might be useful to annotate it here (and of course please double-check that this is correct)

```suggestion
    def _identify_chemically_identical_molecules(self) -> Dict[int, Tuple[int, Dict[int, int]]]:
```

Reply:
Done! And I've given the public method that calls this one an annotation too.